common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java

```diff
@@ -118,76 +119,433 @@ public Collation(
+      private final int collationId;
+
+      private CollationSpecICU(String locale, CaseSensitivity caseSensitivity,
+          AccentSensitivity accentSensitivity) {
+        this.locale = locale;
+        this.caseSensitivity = caseSensitivity;
+        this.accentSensitivity = accentSensitivity;
+        int collationId = ICULocaleToId.get(locale);
+        collationId = SpecifierUtils.setSpecValue(collationId, IMPLEMENTATION_PROVIDER_OFFSET,
+          ImplementationProvider.ICU);
+        collationId = SpecifierUtils.setSpecValue(collationId, CASE_SENSITIVITY_OFFSET,
+          caseSensitivity);
+        collationId = SpecifierUtils.setSpecValue(collationId, ACCENT_SENSITIVITY_OFFSET,
+          accentSensitivity);
+        this.collationId = collationId;
+      }
+
+      private static int collationNameToId(
+          String originalName, String collationName) throws SparkException {
+        // search for the longest locale match because specifiers are designed to be different from
+        // script tag and country code, meaning the only valid locale name match can be
+        // the longest one
+        int lastPos = -1;
+        for (int i = 1; i <= collationName.length(); i++) {
+          String localeName = collationName.substring(0, i);
+          if (ICULocaleMapUppercase.containsKey(localeName)) {
+            lastPos = i;
+          }
+        }
+        if (lastPos == -1) {
+          throw collationInvalidNameException(originalName);
+        } else {
+          String locale = collationName.substring(0, lastPos);
+          int collationId = ICULocaleToId.get(ICULocaleMapUppercase.get(locale));
+
+          // try all combinations of AS/AI and CS/CI
+          CaseSensitivity caseSensitivity;
+          AccentSensitivity accentSensitivity;
+          if (collationName.equals(locale) ||
+              collationName.equals(locale + "_AS") ||
+              collationName.equals(locale + "_CS") ||
+              collationName.equals(locale + "_AS_CS") ||
+              collationName.equals(locale + "_CS_AS")
+          ) {
+            caseSensitivity = CaseSensitivity.CS;
+            accentSensitivity = AccentSensitivity.AS;
+          } else if (collationName.equals(locale + "_CI") ||
+              collationName.equals(locale + "_AS_CI") ||
+              collationName.equals(locale + "_CI_AS")) {
+            caseSensitivity = CaseSensitivity.CI;
+            accentSensitivity = AccentSensitivity.AS;
+          } else if (collationName.equals(locale + "_AI") ||
+              collationName.equals(locale + "_CS_AI") ||
+              collationName.equals(locale + "_AI_CS")) {
+            caseSensitivity = CaseSensitivity.CS;
+            accentSensitivity = AccentSensitivity.AI;
+          } else if (collationName.equals(locale + "_AI_CI") ||
+              collationName.equals(locale + "_CI_AI")) {
+            caseSensitivity = CaseSensitivity.CI;
+            accentSensitivity = AccentSensitivity.AI;
```

Comment:
When we have trimming, do we expect a combinatorial blow up on that we have to test for equality, to do we plan to follow a string substring search approach?

Reply:
It's true that the number of combinations will explode as we add new specifiers for space trimming, punctuation sensitivity, first letter preference etc., and we would need to rework this logic a bit then.

Since the agreement was to scope down this PR only to case/accent sensitivity, I found it simpler to manually specify all combinations and skip adding general-purpose specifier parser logic.